### PR TITLE
8293291: Simplify relocation of native pointers in archive heap

### PR DESCRIPTION
**Pull request by iklam**
Some objects in the archive heap contain native pointers. E.g., archived `java.lang.Class` objects contain a pointer to its `Klass*`.

At runtime, if the archived metadata are mapped at an alternative address, all of the `Klass*` pointers must be relocated. Instead of doing this in an ad-hoc way, this PR uses a bitmap, `FileMapRegion::ptrmap_view()`, to track the position of all the native pointers.

This PR is done in preparation for supporting new types of archived heap objects that have native pointers. E.g., `java.lang.invoke.ResolvedMethodName` has a pointer to a `Method*`.

Notes for reviewers:

- At dump time, the native pointers are remembered in `HeapShared::mark_native_pointers()`.
- At runtime, the native pointers are patched in `ArchiveHeapLoader::patch_native_pointers()`.
- The ad-hoc relocation code has been removed from javaClasses.cpp and systemDictionaryShared.cpp
- Many of the changes are for renaming from "oopmaps" to "bitmaps": the old code handled only a single type of pointers (oop references). The new code handles two types of bitmaps ("oopmap" is for oop references, "ptrmap" is for native pointers).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8293291](https://bugs.openjdk.org/browse/JDK-8293291): Simplify relocation of native pointers in archive heap

### Reviewers
 * [Calvin Cheung](https://openjdk.org/census#ccheung) (@calvinccheung - **Reviewer**) ⚠️ Review applies to [810289f7](https://git.openjdk.org/jdk/pull/10296/files/810289f7b79d36293f435ae6a0949e30bd725494)
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to [810289f7](https://git.openjdk.org/jdk/pull/10296/files/810289f7b79d36293f435ae6a0949e30bd725494)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10296/head:pull/10296` \
`$ git checkout pull/10296`

Update a local copy of the PR: \
`$ git checkout pull/10296` \
`$ git pull https://git.openjdk.org/jdk pull/10296/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10296`

View PR using the GUI difftool: \
`$ git pr show -t 10296`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10296.diff">https://git.openjdk.org/jdk/pull/10296.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/include/cds.h [hunk 1]
```diff
@@ -55,8 +55,12 @@ typedef struct CDSFileMapRegion {
                               // - for heap regions, the base address is the compressed oop encoding base
   size_t  _used;              // Number of bytes actually used by this region (excluding padding bytes added
                               // for alignment purposed.
-  size_t  _oopmap_offset;     // Bitmap for relocating embedded oops (offset from SharedBaseAddress).
+  size_t  _oopmap_offset;     // Bitmap for relocating oop fields in archived heap objects.
+                              // (The base address is the botom of the BM region)
```
Comment (calvinccheung) [inline src/hotspot/share/include/cds.h:59]: typo: botom -> bottom

File: src/hotspot/share/include/cds.h [hunk 2]
```diff
@@ -55,8 +55,12 @@ typedef struct CDSFileMapRegion {
...
   size_t  _oopmap_size_in_bits;
+  size_t  _ptrmap_offset;     // Bitmap for relocating native pointer fields in archived heap objects.
+                              // (The base address is the botom of the BM region).
```
Comment (calvinccheung) [inline src/hotspot/share/include/cds.h:62]: typo: botom -> bottom

File: src/hotspot/share/cds/heapShared.cpp [hunk 3]
```diff
@@ -1796,4 +1825,33 @@ ResourceBitMap HeapShared::calculate_oopmap(MemRegion region) {
...
+    Metadata** p = _native_pointers->at(i);
+    if (start <= p && p < end) {
+      assert(*p != NULL, "must be non-null");
+      num_non_null_ptrs ++;
+      size_t idx = p - start;
+      oopmap.set_bit(idx);
+    }
+  }
+
+  log_info(cds, heap)("calculate_ptrmap: marked %d non-null native pointers our of "
```
Comment (calvinccheung) [inline src/hotspot/share/cds/heapShared.cpp:1848]: typo: our -> out

File: src/hotspot/share/cds/filemap.cpp [hunk 4]
```diff
@@ -1665,8 +1688,9 @@ size_t FileMapInfo::write_heap_regions(GrowableArray<MemRegion>* regions,
     int region_idx = i + first_region_id;
     write_region(region_idx, start, size, false, false);
     if (size > 0) {
-      space_at(region_idx)->init_oopmap(oopmaps->at(i)._offset,
-                                        oopmaps->at(i)._oopmap_size_in_bits);
+      int oopmap_idx = i * 2;
+      int ptrmap_idx = i * 2 + 1;
+      space_at(region_idx)->init_bitmaps(bitmaps->at(i*2), bitmaps->at(i*2 + 1));
```
Comment (calvinccheung) [inline src/hotspot/share/cds/filemap.cpp:1693]: Please add some comment on oopmap_idx and ptrmap_idx.

Line 1693 could be:
`space_at(region_idx)->init_bitmaps(bitmaps->at(oopmap_idx), bitmaps->at(ptrmap_idx));`

File: src/hotspot/share/cds/metaspaceShared.cpp [hunk 5]
```diff
@@ -891,36 +893,55 @@ void VM_PopulateDumpSharedSpace::dump_java_heap_objects(GrowableArray<Klass*>* k
...
+    dump_heap_bitmaps(_open_heap_regions, _open_heap_bitmaps);
   }
 }
 
-void VM_PopulateDumpSharedSpace::dump_heap_oopmaps(GrowableArray<MemRegion>* regions,
-                                                   GrowableArray<ArchiveHeapOopmapInfo>* oopmaps) {
+void VM_PopulateDumpSharedSpace::dump_heap_bitmaps(GrowableArray<MemRegion>* regions,
+                                                   GrowableArray<ArchiveHeapBitmapInfo>* bitmaps) {
   for (int i=0; i<regions->length(); i++) {
-    ResourceBitMap oopmap = HeapShared::calculate_oopmap(regions->at(i));
```
Comment (calvinccheung) [inline src/hotspot/share/cds/metaspaceShared.cpp:907]: Pre-existing: line 908.
While you're touching this code, please add a space before and after the '=' and '<'.

**Review by calvinccheung — COMMENTED** [review]
I've a few minor comments. Thanks.

**Review by calvinccheung — APPROVED** [review]
Updates look good.

**Review by calvinccheung — APPROVED** [review]

**Review by coleenp — APPROVED** [review]
This makes sense and does seem cleaner.

**Comment by iklam** [comment]
Thanks @calvinccheung and @coleenp for the review. I have merged with latest repo (after CPU sync) and retested with tiers 1-4. All passed.
/integrate

**Pull request closed without merging**